### PR TITLE
RUBY-1443 Separate ReplicaSetNoPrimary & ReplicaSetWithPrimary

### PR DESCRIPTION
**Pull request by p-mongo**

### CODE REVIEW

File: lib/mongo/cluster/topology.rb [hunk 1]
```diff
@@ -35,14 +29,26 @@ class Cluster
     # @since 2.0.0
     module Topology
       extend self
+    end
+  end
+end
 
+require 'mongo/cluster/topology/replica_set_no_primary'
```
Comment (saghm) [inline lib/mongo/cluster/topology.rb]: What's the purpose of moving the require statements to the middle of the file?

File: lib/mongo/cluster/topology/replica_set_with_primary.rb [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+module Mongo
+  class Cluster
+    module Topology
+
+      # Defines behavior when a cluster is in replica set topology,
+      # and is a no primary which has been discovered by the driver.
+      #
+      # @since 2.7.0
+      class ReplicaSetWithPrimary < ReplicaSetNoPrimary
```
Comment (saghm) [inline lib/mongo/cluster/topology/replica_set_with_primary.rb]: Just to make sure I understand, are we planning on implementing any behavioral differences between ReplicaSetWithPrimary or ReplicaSetNoPrimary, or is the purpose of the separate classes just to make monitoring easier?

File: spec/spec_tests/data/sdam/rs/secondary_wrong_set_name_with_primary_second.yml [hunk 3]
```diff
@@ -0,0 +1,73 @@
+description: "Secondary wrong setName with primary when secondary responds first"
```
Comment (saghm) [inline spec/spec_tests/data/sdam/rs/secondary_wrong_set_name_with_primary_second.yml]: Has this been merged into the specifications repo yet?

File: lib/mongo/cluster/topology.rb (hunk 1 above)
Comment (p-mongo) [reply lib/mongo/cluster/topology.rb]: I'd move them to the bottom but that does not work.

File: lib/mongo/cluster/topology/replica_set_with_primary.rb (hunk 2 above)
Comment (p-mongo) [reply lib/mongo/cluster/topology/replica_set_with_primary.rb]: The purpose right now is spec compliance which mandates the difference between the two. Depending on how much work/refactoring is done in the other tickets there may be behavior differences. For example, there shouldn't be a need to query servers for primary to find out if a topology has a primary.

File: spec/spec_tests/data/sdam/rs/secondary_wrong_set_name_with_primary_second.yml (hunk 3 above)
Comment (p-mongo) [reply spec/spec_tests/data/sdam/rs/secondary_wrong_set_name_with_primary_second.yml]: This change is not wanted in specs, see my PR there.

File: lib/mongo/cluster/topology.rb (hunk 1 above)
Comment (saghm) [reply lib/mongo/cluster/topology.rb]: Sounds good! Thanks for explaining

File: lib/mongo/cluster/topology/replica_set_with_primary.rb (hunk 2 above)
Comment (saghm) [reply lib/mongo/cluster/topology/replica_set_with_primary.rb]: Makes sense

File: spec/spec_tests/data/sdam/rs/secondary_wrong_set_name_with_primary_second.yml (hunk 3 above)
Comment (saghm) [reply spec/spec_tests/data/sdam/rs/secondary_wrong_set_name_with_primary_second.yml]: Ah, I see Jesse's comment now. Thanks for pointing me to it!

**Pull request merged**